xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.ChannelLogger;
+import io.grpc.ChannelLogger.ChannelLogLevel;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.LoadBalancerRegistry;
+import io.grpc.Status;
+import io.grpc.internal.ObjectPool;
+import io.grpc.internal.ServiceConfigUtil.LbConfig;
+import io.grpc.util.GracefulSwitchLoadBalancer;
+import io.grpc.xds.XdsClient.ClusterUpdate;
+import io.grpc.xds.XdsClient.ClusterWatcher;
+import io.grpc.xds.XdsLoadBalancerProvider.XdsConfig;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * Load balancer for experimental_cds LB policy.
+ */
+public final class CdsLoadBalancer extends LoadBalancer {
+  private final ChannelLogger channelLogger;
+  private final LoadBalancerRegistry lbRegistry;
+  private final GracefulSwitchLoadBalancer switchingLoadBalancer;
+
+  // The following fields become non-null once handleResolvedAddresses() successfully.
+
+  // Most recent XdsConfig.
+  @Nullable
+  private CdsConfig cdsConfig;
+  // Most recent ClusterWatcher.
+  @Nullable
+  private ClusterWatcher clusterWatcher;
+  @Nullable
+  private ObjectPool<XdsClient> xdsClientRef;
+  @Nullable
+  private XdsClient xdsClient;
+
+  CdsLoadBalancer(Helper helper) {
```

Comment:
This is not used anywhere AFAIS : can we remove it or add a comment about why this constructor is there?

Reply:
There should be a `CdsLoadBalancerProvider`.

Reply:
Yes, I added `CdsLoadBalancerProvider` now.